otter/controller.py

```diff
@@ -150,6 +152,91 @@ def obey_config_change(log, transaction_id, config, scaling_group, state,
     return d
 
 
+def delete_group(log, trans_id, group, force):
+    """
+    Delete group based on the kind of tenant
+
+    :param log: Bound logger
+    :param str trans_id: Transaction ID of request doing this
+    :param otter.models.interface.IScalingGroup scaling_group: the scaling
+        group object
+    :param bool force: Should group be deleted even if it has servers?
+
+    :return: Deferred that fires with None
+    :raise: `GroupNotEmptyError` if group is not empty and force=False
+    """
+
+    def check_and_delete(_group, state):
+        if state.desired == 0:
+            d = trigger_convergence_deletion(log, group)
+            return d.addCallback(lambda _: state)
+        else:
+            raise GroupNotEmptyError(group.tenant_id, group.uuid)
+
+    if tenant_is_enabled(group.tenant_id, config_value):
+        return (trigger_convergence_deletion(log, group)
+                if force else group.modify_state(check_and_delete))
```

Comment:
In force case, we are not concerned about current desired and hence trigger convergence deletion since converger will ignore desired. Without force, the check for desired needs to be atomic. Otherwise, one node executing policy could be reading desired 0 and changing it to say 2 returning success and another node deleting could be reading 0 and triggering deletion instead of returning `GroupNotEmptyError`. With the lock, we ensure that deletion and policy execution do not conflict each other.